### DISCUSSION TITLE
Make use of Successive Key Presses

### DISCUSSION THREAD

**Discussion by velitasali** [velitasali/gnome-shell-extension-awesome-tiles#2] [Ideas]
There is something that could be interesting to add on your extension;
i'm talking about the ability to resize to multiple sizes on every window placement.
Let me explain it better:
You put a window on left bottom corner: first iteration of shortcut will place window at 50% of lengh, second iteration will reduce size at 30%, third will set to 70% (percents are an example, it would be gorgeous to set them)

_Originally posted by @Dolland in https://github.com/velitasali/gnome-shell-extension-awesome-tiles/issues/1#issuecomment-1001571549_

**Comment by velitasali** [comment]
@Dolland The idea of making use of successive key presses is good.  Increasing the grid size by every successive press makes sense to me. I think I can add an option to enable this behavior.

**Reply by Dolland** [reply]
Gorgeous! 
(Sorry, i didn't see the "Discussion" tab before :D)

**Comment by velitasali** [comment]
@Dolland I have made some changes that enable 3-grid tiling on this branch here https://github.com/velitasali/gnome-shell-extension-awesome-tiles/tree/successive-tiling. Tbh, I am not satisfied with the results, so your feedback is important here because to me, it is not useful as it is.

**Reply by Dolland** [reply]
Let me test it!

**Reply by Dolland** [reply]
Uhm ok, i think it overall works, i personally would like to customize how it works, let me explain:
i use only corner placement, so for me is not useful that resize affects also height, but i understand it can be useful to who want to place 3 windows on height, so i think it's great how it works but it's mandatory to let users customize the behaviour

**Reply by velitasali** [reply]
Ok, with the last commit, it works as you first described it. First press does 2/1 tiling, second 3/2, and third 3/1. Let me know what you think. This might also fix #3. 

Btw, thanks for helping.

**Reply by Dolland** [reply]
ok i see it also changes height, which is not what i'm searching for but i understand it makes sense.
For me it lacks only adjustable preferences on how and how much resize width/height and then it's almost perfect :D 
Also don't forget to put "delete keybinding" option, and BIG thanks to your work!

P.S.
Another implementation (but i think it maybe better to make it optional so we can choose which behavior user wants) could be to move windows not absolutely but relatively to actual position.
Let me explain:
You are on bottom left corner, hitting the "up" keybind doesn't put window on absolute top of the screen but on top relatively to the bottom left corner.
So if you have bottom-center-top and you are on bottom left, hitting the "top" keybind puts window on center left, another hit puts window on top left, "right" keybind puts windows on top center and so on.

I understand it's a totally different behavior from how the extension works by now, but it would be interesting to can have dual functionality :)

**Reply by Dolland** [reply]
ah! in this week i think i'll work on traslating the extension in Italian :)

**Reply by velitasali** [reply]
> ok i see it also changes height, which is not what i'm searching for but i understand it makes sense. 

I am not sure what do you mean by height. Can you elaborate? 

> For me it lacks only adjustable preferences on how and how much resize width/height and then it's almost perfect 

Yes, it is not an option yet. It will need planning first :)


> P.S. Another implementation (but i think it maybe better to make it optional so we can choose which behavior user wants) could be to move windows not absolutely but relatively to actual position. Let me explain: You are on bottom left corner, hitting the "up" keybind doesn't put window on absolute top of the screen but on top relatively to the bottom left corner. So if you have bottom-center-top and you are on bottom left, hitting the "top" keybind puts window on center left, another hit puts window on top left, "right" keybind puts windows on top center and so on.

This will need a grid system. Right now, there are other things to be decided first. First one being whether the column and row sizes should be dynamic. To me, keeping it to 3x3 at most makes perfect sense because I don't have a very 40"+ monitor. So, I'd rather finish this feature first :)

> ah! in this week i think i'll work on traslating the extension in Italian :)

That would be great :)

**Reply by Dolland** [reply]
By "height" i mean that if i cycle the "top right" position, it changes not only how much "large" is the windows but also how much "tall" is it.
I'm sure this behavior is legit because this way we can put top+center+bottom windows when on smaller resizing, but i'd also like the possibility to resize only on the horizontal size.
Hope this can be a better explanation, i'm not native english so... :D

 as for my implementation idea, ok, i suppose it needs a bit of work under the hood so perhaps it's better to refine what we already have :)

**Reply by velitasali** [reply]
> By "height" i mean that if i cycle the "top right" position, it changes not only how much "large" is the windows but also how much "tall" is it.
> I'm sure this behavior is legit because this way we can put top+center+bottom windows when on smaller resizing, but i'd also like the possibility to resize only on the horizontal size.

`<Super> Keypad 2, 4, 6, 8` keys do that already. I don't understand the difference.

**Reply by Dolland** [reply]
ok, let's see if i can explain it better:

Here is the code that (i think) does all stuffs to resize and put window:

` 

    const area = this._calculateArea(window)
    const prev = this._previousTilingOperation
    const windowId = window.get_id()
    const successive = prev && prev.id == windowId && prev.t == top && prev.b == bottom
      && prev.l == left && prev.r == right
    const secondIteration = prev && prev.second

    const gridSize = successive ? 3 : 2
    const gridUse = !successive || secondIteration ? 1 : 2

    const w = left && right ? area.w : Math.round(area.w / gridSize) * gridUse
    const h = top && bottom ? area.h : Math.round(area.h / gridSize) * gridUse
    const x = area.x + (left ? 0 : area.w - w)
    const y = area.y + (top ? 0 : area.h - h)`

the third last line, i keep supposing, adjust the height of the window according to the 3 sizes that i can choose, e.g., on top right corner.
So in fact the height is also changed when i cycle the 3 sizes on that corner, isn't it?
Ok, if it's all correct, it would be nice to have a option to **disable** this height adjustment.

I'm sorry if that's all wrong about the code, i'm not an expert at all :)

**Reply by velitasali** [reply]
It does change the height, but it covers the whole vertical space, leaving only the user defined gaps. If you mean leaving it unaltered, then yes, it is not possible :)

**Reply by Dolland** [reply]
Yes yes, it's what I'm talking about!
Ok, that's a bad news for me :-D
But I understand :-)

**Reply by velitasali** [reply]
Alright, then I am going to move these changes to the main branch after I get feedback from #3. 

Thank you for the help :)

**Reply by x9sim9** [reply]
How about something as simple as holding the win + shift key, so if you press any key and shift it makes the size half of what it would normally be.

One Press -> win + up 50%, win + shift + up 25%
Two Press -> win + up 70%, win + shift + up 35%
Three Press -> win + up 30%, win + shift + up 15%

I understand you are trying to keep it simple and not break the existing simplicity

**Reply by x9sim9** [reply]
The other idea i thought was just allowing a use to choose what the snaps are with a simple options panel.

Snap 1 = 50%
Snap 2 = 70%
Snap 3  = 25%

and a use can just change these values to whatever they prefer...
